TriblerGUI/tribler_request_manager.py

```diff
@@ -37,104 +37,123 @@ def __init__(self, max_outstanding=50, timeout=15):
+         - MEDIUM
+         - LOW
 
         :param request_manager: the TriblerRequestManager wishing to perform a request.
-        :param callback: the callback to call if when this request is allowed to be sent.
+        :param method: request method.
+        :param endpoint: request endpoint.
+        :param data: request data.
+        :param read_callback: callback to call if the request is processed.
+        :param capture_errors: whether to display the errors or not.
         :param priority: the priority for this request.
         """
-        # Send CRITICAL requests immediately
-        if priority == QueuePriorityEnum.CRITICAL:
-            callback()
-            return
         self.lock.acquire()
-        queue_length = len(self.queue)
-        insert_point = -1
-        insert_guess = 0
-        insert_step = 1
-        insert_find = QueuePriorityEnum.CRITICAL
-        # Create an educated guess to find where in the queue to insert a request
-        # If the queue is too full, drop/cancel the request
-        if priority == QueuePriorityEnum.HIGH and queue_length > 0:
-            insert_guess = max(min(int(queue_length * self.old_medium_index), queue_length), 0)
-            insert_step = 1 if self.queue[insert_guess] == QueuePriorityEnum.HIGH else -1
-            insert_find = QueuePriorityEnum.MEDIUM if insert_step == 1 else QueuePriorityEnum.HIGH
-        elif priority == QueuePriorityEnum.MEDIUM and queue_length > 0:
-            if queue_length > self.max_outstanding:
-                logging.error("Too many requests to handle new request with priority %s.", priority)
-                request_manager.cancel_request()
-                return
-            insert_guess = max(min(int(queue_length * self.old_low_index), queue_length), 0)
-            insert_step = 1 if self.queue[insert_guess] == QueuePriorityEnum.MEDIUM else -1
-            insert_find = QueuePriorityEnum.LOW if insert_step == 1 else QueuePriorityEnum.MEDIUM
-        else:
-            if queue_length > self.max_low_outstanding:
-                logging.error("Too many requests to handle new request with priority %s.", priority)
-                request_manager.cancel_request()
-                return
-            insert_point = queue_length
-        # Find the point to insert the request in the queue
-        if insert_point == -1:
-            for i in xrange(insert_guess, queue_length if insert_step == 1 else 0, insert_step):
-                if self.queue[i] == insert_find:
-                    insert_point = i + (0 if insert_step == 1 else 1)
-                    break
-        # Update pointers
+        queue_item = (request_manager, endpoint, read_callback, data, method, capture_errors, time())
+        if priority == QueuePriorityEnum.CRITICAL:
+            if len(self.critical_queue) < self.max_outstanding:
+                self.critical_queue.append(queue_item)
+            else:
+                # no space left, reduce the priority to high
```

Comment:
Why not just boot the lowest priority request and add this to the queue as CRITICAL priority?

Reply:
Yes, now lowest priority request is booted out if no space is left in other queues.